proof/refine/ARM/Reply_R.thy

```diff
@@ -1081,27 +1024,841 @@ lemma sc_replies_relation_replyNext_update:
+     od)"
+  apply (clarsimp simp: monadic_rewrite_def)
+  apply (rule monad_state_eqI)
+    apply (find_goal \<open>match conclusion in "_ = _" \<Rightarrow> -\<close>)
+    subgoal
+    apply (insert no_fail_updateReply, drule_tac x=rp and y="replyPrev_update Map.empty" in meta_spec2)
+    apply (insert no_fail_updateReply, drule_tac x=rp and y="replyNext_update Map.empty" in meta_spec2)
+    apply (rule; clarsimp simp: in_monad no_fail_def snd_bind split_def)
+     apply (drule (1) use_valid[OF _ updateReply_reply_at'], fastforce)+
+    done
+   apply (all \<open>clarsimp simp: updateReply_def getReply_def setReply_def getObject_def2
+                              obj_at'_def projectKOs updateObject_default_def setObject_def
+                              loadObject_default_def2[simplified] ARM_H.fromPPtr_def ps_clear_upd'
+                              split_def in_monad in_magnitude_check' objBits_simps'\<close>)
+   apply (fastforce simp add: fun_upd_def replyPrevNext_update_commute)+
+  done
+
+lemma reply_sc_update_sc_with_reply_None:
+  "set_reply_obj_ref reply_sc_update rp None \<lbrace>\<lambda>s. sc_with_reply rp s = None\<rbrace>"
+  apply (wpsimp wp: update_sk_obj_ref_wp)
+  apply (drule sc_with_reply_NoneD)
+  apply (simp add: sc_with_reply_def sc_replies_sc_at_def obj_at_def)
+  apply (rule the_pred_option_None)
+  by (metis Structures_A.kernel_object.simps(45) option.simps(1))
+
+(* cleanReply corres when rp is not in a reply stack *)
+lemma cleanReply_sc_with_reply_None_corres:
+  "corres dc (reply_at rp and pspace_aligned and pspace_distinct
+    and valid_replies and (\<lambda>s. sc_with_reply rp s = None))
+    \<top>
+     (set_reply_obj_ref reply_sc_update rp None)
+     (cleanReply rp)"
+  apply (rule_tac Q="reply_at' rp" in corres_cross_add_guard)
+   apply (fastforce dest!: state_relationD reply_at_cross)
+  apply (simp add: cleanReply_def bind_assoc)
+  apply (rule corres_guard_imp)
+    apply (rule monadic_rewrite_corres'[OF _ updateReply_Prev_Next_rewrite])
+    apply (rule corres_bind_return)
+    apply (rule corres_guard_imp)
+      apply (rule corres_split[OF updateReply_replyNext_not_head_corres])
+         apply (simp add: isHead_def)
+        apply (rule_tac P="reply_at rp and valid_replies and (\<lambda>s. sc_with_reply rp s = None)"
+                    and P'="reply_at' rp" in corres_noop_sr)
+          apply (rule sr_inv_imp)
+            apply (rule updateReply_sr_inv_prev[simplified])
+           apply (clarsimp simp: reply_relation_def)
+          apply simp
+         apply (wpsimp wp: reply_sc_update_sc_with_reply_None simp: isHead_def)+
+     apply simp+
+  done
+
+(* corres between update_sched_context and updateReply *)
+lemma updateReply_replyPrev_takeWhile_middle_corres:
+  assumes
+    "rp \<in> set (sc_replies sc)"
+    "st = Structures_A.thread_state.BlockedOnReply rp"
+    "st' = Structures_H.thread_state.BlockedOnReply (Some rp)" "hd (sc_replies sc) \<noteq> rp"
```

Comment:
* It appears that `st'` is not used?
* If `st = ...` is in the assumptions, I would prefer for it to be substituted into the preconditions. Otherwise this corres rule will produce one more subgoal than is needed, and one more schematic than is needed.

Reply:
Thanks, it turned out that all references to `t` are unnecessary in this lemma. So, `st_tcb_at` itself can go.